src/base/LruMap.h

```diff
@@ -148,50 +184,70 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
-    storage.insert(MapPair(key, index.begin()));
+    // optimization: save on memory usage for already-stale objects
+    // (findEntry() would discard instead later anyway)
+    if (ttl < 0)
+        return false;
+
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memoryUsed();
+    index.push_front(e);
+
+    auto m = MapPair(key, index.begin());
+    memUsed_ += memoryUsedBy(m);
+    storage.insert(m);
 
     ++entries_;
-    return true;
+
+    // if over memory limit, start purging
+    trim();
```

Comment:
I think it is wrong to add entries that we are going to purge immediately (and all other perfectly valid entries with them!). We should use the same principle we use for prematurely expired entries: If the entry cannot be found later, do not store it now. Here is a sketch:

```
std::unique_ptr<Entry> freshEntry(new Entry(key, t, ttl));
const auto extraMemory = freshEntry->memoryUsed();
if (extraMemory > memLimit())
    return false; // does not fit

index.push_front(freshEntry.release());
storage.insert(MapPair(key, index.begin()));
memUsed_ += extraMemory;
++entries_;

trim();
assert(entries_);
return true; // because LRU will not remove the freshest/last entry
```

Please note that the sketch does not call memoryUsedBy(). An equivalent call can be used here, but I think we should simplify by allowing Entry::memoryUsed() to account for _all_ memory used by the entry payload in LruMap, including the duplicated key storage. AFAICT, the updated Entry::memoryUsed() method does not need to account for sizeof(MapPair) IMO because sizeof(MapPair) reflects the size of small metadata, not payload, but you can adjust Entry::memoryUsed() to account for sizeof(MapPair) if needed.

Using `unique_ptr` is good style (because it stops leaks during exceptions), but you can remove it by refactoring Entry::memoryUsed() in to a static method with parameters.

Reply:
That case is quite rare, most of these caches are hundreds of KB. Now that Key is an SBuf addition is extremely cheap too. The far more common case is to leave the new entry and purge old ones.

Also, If we start eliding things because they might push others out of the cache where do we stop ... anything that adds a single byte of memory usage, or anything expiring in the next 64 years? IMO it is far better to consistently do the same thing of rolling new items in and old ones out.

Reply:
> That case is quite rare

... but still should be handled correctly, especially since it is easy to handle it correctly. Besides, the frequency of that case depends on the configuration and environment/usage. It may be a frequent case somewhere.

> If we start eliding things because they might push others out of the cache where do we stop

I am not suggesting to prohibit caching of entries that "might push others out"! I insist that we prohibit caching of entries that cannot be cached. Such entries should have no effect on the cache.

Reply:
The map size is 256KB for most helpers and 2MB or 4MB in the case of SSL helpers. Admin tend to configure it upwards from that or disable completely. The maximum helper response size is hard coded to 64KB and tends to be smaller than 8KB except for Kerberos helpers.

So no entry comes close to exceeding the map size without admin intervention / misconfiguration. I do not believe the extra cost (1.5x if-statements and 2x entry size calculations) checking this on every addition is worth the price given the rarity of encountering it.

Reply:
I do not think correct code requires significantly more checks and calculations in this case. If I am right, than we should write correct code.